bus-mapping/src/rpc.rs

```diff
@@ -121,15 +141,31 @@ impl<P: JsonRpcClient> GethClient<P> {
         Ok(resp.0.into_iter().map(|step| step.result).collect())
     }
 
-    /// Calls `eth_getProof` via JSON-RPC returning a [`EIP1186ProofResponse`]
-    /// returning the account and storage-values of the specified
-    /// account including the Merkle-proof.
+    /// Calls `eth_getCode` via JSON-RPC returning a contract code
+    pub async fn get_code_by_address(
+        &self,
+        contract_address: Address,
+    ) -> Result<Vec<u8>, Error> {
+        let address = serialize(&contract_address);
+        let id = serialize(&"0x2".to_string());
+        let code: Bytes = self
+            .0
+            .request("eth_getCode", [address, id])
+            .await
+            .map_err(|e| Error::JSONRpcError(e.into()))
+            .unwrap();
+        Ok(code.to_vec())
+    }
+
+    /// Calls `eth_getProof` via JSON-RPC returning a
+    /// [`EIP1186CustomProofResponse`] returning the account and
+    /// storage-values of the specified account including the Merkle-proof.
     pub async fn get_proof(
         &self,
         account: Address,
         keys: Vec<Word>,
         block_num: BlockNumber,
-    ) -> Result<EIP1186ProofResponse, Error> {
+    ) -> Result<EIP1186CustomProofResponse, Error> {
```

Comment:
That seems wrong to me.
You're basically returning the response as `EIP1186CustomProofResponse` to the user.
When you could use internally the  `EIP1186CustomProofResponse` to deserialize the response data, and then impl a conversion between this type and the one in `ethers` types (`EIP1186ProofResponse`).

I had to do the same here: https://github.com/appliedzkp/zkevm-circuits/blob/8aa831e75dc98cdd2febb1645cb16f9071b905ba/bus-mapping/src/rpc.rs#L100-L106

You can check the two types and you'll understand what I mean :smile:

Reply:
I noticed that `https://docs.rs/ethers-core/0.6.1/ethers_core/types/struct.EIP1186ProofResponse.html` only contains private fields, which makes it kind of useless?  I think this is something that needs to be addressed in `ethers`, but meanwhile I think the best option is to use an internal declaration of `EIP1186ProofResponse` with public fields.

So I would do the following:
1. Don't import `EIP1186ProofResponse` from `ethers` anywhere
2. In `EIP1186CustomProofResponse` make the fields `pub`
3. Rename `EIP1186CustomProofResponse` to `EIP1186ProofResponse`

Reply:
> I had to do the same here:

Thank you for the polite explanation!
> So I would do the following:

I understand.